### ISSUE TITLE
List of Bugs/Requests/Things

### ISSUE THREAD

**Issue by 540howdy** [HeisenBugDev/QuantumCraft#155]
So far the mod is pretty fun
### Bugs
- ~~Tried right clicking one of the quantumcraft wires with your special wrench and it crashed Log: http://pastebin.com/gfBcE3UQ~~
- ~~Generator is missing the fancy green charging bar. It's empty, but it seems reasonable that once the generator is full this bar should also be full.~~
- ~~Wires take forever to take with picks, unless they're supposed to be broken by the wrench. (this crashes as stated above)~~
- ~~Build 344: Can't open the GUI of any blocks.. Tried running it on a 100+ modpack then tried running it only with NEI. Didn't work on either~~ Fixed in later builds
- ~~The Quantum De-Energizer never fills up completely with energy. I think this is because of the amount of energy each crystal has.~~ Doesn't matter
- ~~Just an opinion, but the Energy Extractor should be a built in GUI on a generator (Quantum De-energizer) instead of it being on its own block~~ We thought about it when we implemented it, the extractor is meant for items that are processed and ready to use whereas things like the Deenergizer are for processing the materials themselves.
- ~~The quantum de-energizer on the Getting Started page does not have a hyperlink to its article like the other items~~ Fixed
### Recommendations
- The ION tunneler is pretty unclear when it's mining. Maybe a laser or a visual to show that's it doing something?
- The ION Scanner should also receive a visual.
- It seems I am able to run as many machines as I want on a single power line from a single generator with 0 energy decrease in efficiency. If this is right, may want to tweak it a bit.
- Energy comes in the form of an ore found in the ground. This is your chance to create an energy system unique from other mods, one that's renewable.
- In-game guide.
## The Big Question™

**_Why is this better than any other tech mod?_**

The theme of it is not complete and it doesn't add machines that completely make your game experience more enjoyable yet (at least not as much any ThermalExp) but it is adding some cool things that should catch on (IN WORLD GUIs!)

Okay I think I'm done, nice mod so far looking forward to seeing it in the future.
Thanks

**Comment by hunterboerner** [comment]
For the generator, do you mean steam gen?

**Comment by hunterboerner** [comment]
@540howdy, I just pushed a fix, can you test with the latest build?

**Comment by 540howdy** [comment]
@hunterboerner Testing out 344 right now

**Comment by hunterboerner** [comment]
@540howdy status update?

**Comment by 540howdy** [comment]
@hunterboerner Oh, hah I updated the original post. None of the GUI's are opening for me on a NEI + Quantumcraft world nor on a 100+modpack world

**Comment by sammko** [comment]
Can I be informed why we have a te variable on L50 and another one on L54 just initialized in a different way?

**Comment by hunterboerner** [comment]
Err.. it shouldn't make a difference but the second will return null if it's NOT an instance of TileMachineBase which will prevent casting errors if we happen to init a non TileMachineBase te in the block. <= I hope that makes sense

**Comment by sammko** [comment]
Yes it does make sense, but then why don't you get rid of the one on L50 and replace it w/ the one from L54 decreasing RAM usage and making it generally more efficient

**Comment by hunterboerner** [comment]
https://github.com/HeisenBugDev/QuantumCraft/blob/feature/146-multipart/src/main/java/quantumcraft/blocks/abstractblocks/BlockMachine.java#L74

If I do that then we can't have a tile that is IUpgradable but NOT TileMachineBase. I could do a check to see if te is instance of TileMachineBase but that may not be much more efficent.

**Comment by sammko** [comment]
ugh... that onBlockActivated method is like the worst thing in our entire mod... and its broken 50% of the time

**Comment by hunterboerner** [comment]
@sammko, hasn't broken for me.

**Comment by hunterboerner** [comment]
@540howdy, do you know what build you first tested on?

**Comment by hunterboerner** [comment]
@540howdy, I found the problem. By changing the modid to get it to work with qCraft, it broke guis and a few other things so reverting and no qCraft + quantumcraft for a while.

**Comment by 540howdy** [comment]
@hunterboerner 344, didn't realize there was a 345 out. So 344 has the broken GUI bug, but 345 fixes it. I'll look for more bugs/things in this patch then.

**Comment by hunterboerner** [comment]
@540howdy, yeah, just released that. You won't be able to run qCraft and QuantumCraft though.

EDIT: Thanks for the help.

**Comment by 540howdy** [comment]
No problem. Hmm

The Quantum-Multitool works now, can shift click wires to break. Nice 

EDIT by @hunterboerner: Moved thing to the OP.

**Comment by hunterboerner** [comment]
@540howdy, for number 3: It is. The Capacitors have Inject/Extract slots and the reason we kept those two machines in is for automation.

**Comment by hunterboerner** [comment]
@540howdy, do you have any ideas for new features that would be creative?

**Comment by 540howdy** [comment]
@hunterboerner I feel like the best step forward for this mod is to have someone (possibly you) research into Quantum Mechanics and revolve the mod around the theory of Quantum Mechanics. Qcraft takes this idea in a very simple form, and does not offer a tier of tech related ideas related to Quantum Mechanics. So far, Thermal Expansion offers a very sleek way of playing. It's an easy to climb mod. On the other hand, Industrialcraft seems to force tedious recipes that can be easily solved with factories or automation. Your mod however can offer a new way of doing tech with the vast idea that is Quantum Mechanics. I have very little background on this, except for the very basic ideas. Good luck.

**Comment by sammko** [comment]
I actually came up with this mod, and about the research into Quantum mechanics, the dislocator is based on the principle of Quantum dislocation / entanglement. Entanglement originally works with eg. polarization of photons, but i mirrored it on blocks in minecraft.

**Comment by sammko** [comment]
Next, thanks to @LordFokas, we are planning to implement the observer effect ( #82 ) and String theory ( #83 )

**Comment by sammko** [comment]
Also quantum tunnelling is reserved for the Quantum Wormhole Inducer ( #9 )

**Comment by hunterboerner** [comment]
Any ideas (less vague) for something you feel is missing?

**Comment by 540howdy** [comment]
@sammko That goes to show you how little I know about this. A bunch of mods have a "Getting Started" guidebook. Maybe one for this mod too that briefly explains quantum mechanics and what each block does. I was talking more about the generator/furnace. Smelting a crystal found in the ground that then can be used for fuel to power a furnace is kinda odd, it feels out of place. But the tunneling is pretty cool, only thing that would make it more awesome would be a visual.

**Comment by hunterboerner** [comment]
@540howdy, you mean the Getting Started page? http://heisenbugdev.com/wiki/a_meta/guide

**Comment by sammko** [comment]
Yeah we could implement that in-game

**Comment by sammko** [comment]
We should also somehow get superposition going

**Comment by 540howdy** [comment]
@hunterboerner Yeah but in game. It's a simple feature, but could go a long way. + a bit of information about quantum mechanics

**Comment by sammko** [comment]
The dislocator is actually mis-named, we should rename it to entangler or something becuase back then i had quantum nonlocality and entanglement messed up

**Comment by sammko** [comment]
This is a list of Quantum fundamental concepts (thanks to wikipedia)
-    Complementarity - no idea what it is
-    Decoherence - no idea what it is
-    Entanglement - Dislocator (should be renamed)
-    Nonlocality - doing research ATM
-    Quantum state - no idea how to implement
-    Superposition - planning
-    Tunnelling - Wormhole inducer
-    Uncertainty - No idea how to implement in-game
-    Wave function - no idea how to implement in-game
-    Symmetry - still doing research
-    Measurement - This isnt really a phenomenon

I will talk to my friend, he also reads a lot about this stuff, but he doesnt do java (yet :P) maybe we could figure out an implementation for some of the phenomenons

Anyways, thanks for the feedback :)

**Comment by hunterboerner** [comment]
@sammko, tell me how you think it'll work and I'll try to code it up.

**Comment by sammko** [comment]
Well idk yet, and I didnt say i wont code in the future, i enjoy coding, but i just wasnt in the mood.. i feel like i will do more coding after holidays. will see what happens

**Comment by hunterboerner** [comment]
I know, but I need something to code :). I asked @MrTJP (ProjectRed) about multiparts and I _think_ he will turn the  qwire into one for us.

**Comment by sammko** [comment]
oh cool

**Comment by hunterboerner** [comment]
> oh cool

btw, I just figured out you can select text and hit `r` on github to quote it!

**Comment by sammko** [comment]
also cool

**Comment by hunterboerner** [comment]
> Energy comes in the form of an ore found in the ground.. This is your chance to create an energy system unique from other mods.. one that's renewable.

Anyone got any ideas? I started steam but I don't think that's the way to go. @540howdy, what's in your brain?

**Comment by 540howdy** [comment]
@hunterboerner Oh gosh. I agree with steam not being the right path for this mod.

Can the Quantum Theory even be used for energy? I mean it can in the simplest of terms, but can it be used to describe a very specific instance?

**Comment by hunterboerner** [comment]
> Can the Quantum Theory even be used for energy? I mean it can in the simplest of terms, but can it be used to describe a very specific instance?

Take a look at #162

**Comment by 540howdy** [comment]
How about a few different furnaces that each apply a different theory of quantum mechanics?
- Quantum Interdimensional Generator: Applies the Theory of Superposition (Copied from #162)
- Insert Name Here: Applies the Theory of Entanglement: My mind keeps darting back to robots. Heh. It would be cool to have little robots that keep a furnace in a constant state. A furnace that only runs when a robot is operating it, and the robot would operate on some sort of fuel.
- Insert New Name Here: Applies the Theory of Decoherence, I found a cool article here: http://www.ipod.org.uk/reality/reality_decoherence.asp. For decoherence, something about rainbows. My creativity took a dive here.
- Insert Another Name Here: Applies the Theory of Heisenburg's Uncertainty Principle: This principle seems to touch on the uncertainty of life. Hah that's in the name. However this can be applied to a furnace, possibly. How do we know that a furnace is always running at exactly 5 ticks of energy per second? Exactly, we don't. You can make a furnace with upgrades to force it to run at a certain amount of energy. Default is random, 1 energy, 2, 9, 12, 15, per second. With upgrades, you can force it to run at 15, 30, 60, etc. Users can choose to keep it at random or force upgrades upon it. This could be the final tier.

Edit: Sorry for the late responses, I"m doing research on google trying to wrap my mind around this insanity.

**Comment by 540howdy** [comment]
@hunterboerner I forgot to mention what the furnaces would run on.. I think the crystals from your original idea could still work. However, the crystals would function differently per furnace. 

Side Note: These are all ideas, I don't want to impose too much.

**Comment by LordFokas** [comment]
You could use the uncertainty to make a furnace that operates on the same principle. You can't know if it's done or not before you open the GUI and look. It won't start working until you close the GUI. the more you wait, the higher the chance more items will be done when you look, but while you look, it completely freezes.

So right there, you have the theories of uncertainty and observer effect.

**Comment by hunterboerner** [comment]
And until we ask @LordFokas if he has any good ideas, we won't know for certain... :P

EDIT: Incorrect. We will know it's one of 4 states which all equal yes in layman's terms.

**Comment by hunterboerner** [comment]
Ok, let's begin extracting the features we want to focus on into their own tickets to organize them and track.

> Furnace 1: 
> Quantum Interdimensional Generator: Applies the Theory of Superposition (Copied from 162)

My question on that is: Why are we generating power in a furnace? This is power gen!

> It would be cool to have little robots that keep a furnace in a constant state

It couldn't just be "robots" because those would be affected by the states too considering they can't observe it directly (ya know, they're robots). However, if you were somehow to be observing this said furnace via the robot, would that be good enough? @sammko, you're the Quantum Expert so what do you think? Maybe a machine that accepts the data feeds from all the robots and allows you to monitor the status of each through a HUD. That way, you are constantly viewing it so the quantum mechanics still apply and you can automate it.

> Furnace 2

The robot idea can be applied to all furnaces. No need for another.

> For decoherence, something about rainbows. My creativity took a dive here.

No idea what you are talking about.

> This could be the final tier.

Should we have all the furnaces work on an upgrade system? Not just a right click on machine one but say an "upgrader" machine that allows automation of upgrading of machines rather than tiered crafting recipes.

> You can make a furnace with upgrades to force it to run at a certain amount of energy. Default is random, 1 energy, 2...

The problem with "random" is that it's like flipping a coin, if you flip it 100 times and say head = 0 and tails = 1, you're going to get about 50 no matter what. We can't say in the period of 1 second use this power, we would have to do something say, in the period of the current itemstack so it will work on a larger scale which won't be able to level out as easily.

> This is a list of Quantum fundamental concepts

@sammko, can you fill in that list (or respond with the info and let me organize it) so that we can have an easy reference for maturing ideas and not sounding like Quantum Idiots :+1: 

> thanks for the feedback :)

If anyone know other people (what people, who are they?), could you ask them to bring in the feedback and ideas. Could someone setup a poll (1-3 questions only, multiple choice) which should be very quick and easy to complete that will provide us a lot of information about what people think we should do.

### Question 1

Current plans, which one first

### Q2

Things we are thinking about, which best

### Q3

Which aspect of MC should we improve (Material Collection, Material Processing, Material Prettyness, Mobs...)

**Comment by hunterboerner** [comment]
I was thinking, what about TE3 style side switching on machines?

**Comment by sammko** [comment]
Yeah I had that idea too. It would be a PITA to implement. But we could try
what OpenBlocks have cause its opensource and even cooler than TE's

**Comment by hunterboerner** [comment]
@sammko, it's in here: https://github.com/OpenMods/OpenModsLib/tree/master/src/openmods/gui. Looks like you're going to have fun :+1:

**Comment by hunterboerner** [comment]
So guys, here's a very early concept image for showing what blocks are scheduled to be broken.
![2013-12-25_13 25 27](https://f.cloud.github.com/assets/1314119/1809014/a4ce8b6e-6d9a-11e3-81c7-fad35aa0655b.png)

**Comment by hunterboerner** [comment]
BTW, the reason you have to be touching _a_ block is because it's DrawBlockHighlightEvent, so if someone knows of another event that just draws the blocks, then we could use that.

**Comment by hunterboerner** [comment]
![2013-12-26_12 29 06](https://f.cloud.github.com/assets/1314119/1811601/ab7036a4-6e5b-11e3-83a1-d30fc5fec87e.png)

Look awesome? Warning: The red blocks which is what the tunneler is mining is very laggy so we may want to have that only visible if you are wearing something.

**Comment by hunterboerner** [comment]
Ok, just noticed that if you break the tunneler, the block thingies won't go away. Trying to fix, not working. This is very laggy anyway so I think I'm going to remove it for now (just disable) because it doesn't look or work good.

**Comment by hunterboerner** [comment]
@540howdy, try build 366 http://heisenbugdev.com/downloads. @sammko, you may also like the new in world guis. Just put your crosshairs on a machine (ION Scanner is most interesting). What do yer think? :rabbit2:

**Comment by sammko** [comment]
In-world GUIS? sounds cool! I'm probably gonna go home tomorrow or the day
after so I will check it out then

**Comment by hunterboerner** [comment]
![2013-12-26_17 35 53](https://f.cloud.github.com/assets/1314119/1812637/9843bb0c-6e86-11e3-91cc-d0982687ce28.png)

Mmmmmm. Look at that screenshot!

**Comment by sammko** [comment]
Oh that's so cool

**Comment by hunterboerner** [comment]
Yeah, the idea came from RS485/Logisticspipes which has something like that but it's more of a hidden feature (theirs is MUCH more complex, including buttonz). This is always there rather than being really hard to find.

**Comment by hunterboerner** [comment]
> The ION tunneler is pretty unclear when it's mining. Maybe a laser or a visual to show that's it doing something?

@540howdy, does the latest build fix this? If you look at the block it gives you info (blocks left...)

> The ION Scanner should also receive a visual.

Maybe a scanning blue laser coming from it? @sammko gets this one.

> It seems I am able to run as many machines as I want on a single power line from a single generator with 0 energy decrease in efficiency. If this is right, may want to tweak it a bit

Don't have any ideas on a unique way to solve this. Don't think it's really a _problem_

> Energy comes in the form of an ore found in the ground. This is your chance to create an energy system unique from other mods, one that's renewable.

Any more ideas other than the current #162 (QIG)?

BTW @540howdy, I edited your original post to add formatting to it. Visit here if you need any more help with it: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet or ask me.

**Comment by LordFokas** [comment]
you can limit how much power each power source can emit every tick.

**Comment by hunterboerner** [comment]
@LordFokas, I don't think that's a good way to handle it. Things like the capacitor **need** to be giving all it's got.

**Comment by hunterboerner** [comment]
Finished the QIG guys.

**Comment by hunterboerner** [comment]
Bug with guis:

![2013-12-29_16 40 24](https://f.cloud.github.com/assets/1314119/1820237/d9325144-70db-11e3-84a9-ab8d598a63d0.png)
